### PR TITLE
DISCOVERYACCESS-8172 - virtual browse - guardrails for bots

### PR DESCRIPTION
**Pull request by bmckinney**
- return appropriate response codes when bots call virtual browse endpoints with missing or incorrect headers and/or params

- [DISCOVERYACCESS-8172](https://culibrary.atlassian.net/browse/DISCOVERYACCESS-8172)

- also addresses [DISCOVERYACCESS-8180](https://culibrary.atlassian.net/browse/DISCOVERYACCESS-8180)

- the rspec controller test might be a bit overkill?

- to me, the following set of curl calls is a better approach to testing the response codes short of creating rspec request tests:

```shell
# no xhr request, no params = 404 Not Found
# thrown by routes.rb - lambda { |req| req.xhr? && req.format == :js }
# previous: 406, UnknownFormat
echo "TEST [1] no xhr request, no params = 404 Not Found"
curl -I 'http://0.0.0.0:9292/get_next'
curl -I 'http://0.0.0.0:9292/get_previous'
curl -I 'http://0.0.0.0:9292/get_carousel'

# xhr request, no params = 400 Bad Request
# thrown by: virtual_browse.rb - params.require([:callnum, :start])
# previous: 406, UnknownFormat
echo "TEST [2] xhr request, no params = 400 Bad Request"
curl -I 'http://0.0.0.0:9292/get_next' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'
curl -I 'http://0.0.0.0:9292/get_previous' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'
curl -I 'http://0.0.0.0:9292/get_carousel' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'

# xhr request, params, accept header doesn't include javascript = 404 Not Found
# thrown by routes.rb - lambda { |req| req.xhr? && req.format == :js }
# previous: 406, UnknownFormat
echo "TEST [3] xhr request, params, accept header doesn't include javascript = 404 Not Found"
curl -I 'http://0.0.0.0:9292/get_next?callnum=HB172.5%20.F692%202019&start=1' -H 'Accept: text/html' -H 'X-Requested-With: XMLHttpRequest'
curl -I 'http://0.0.0.0:9292/get_previous?callnum=HB172.5%20.F692%202019&start=1' -H 'Accept: text/html' -H 'X-Requested-With: XMLHttpRequest'
curl -I 'http://0.0.0.0:9292/get_carousel?callnum=HB172.5%20.F692%202019' -H 'Accept: text/html' -H 'X-Requested-With: XMLHttpRequest'

# xhr request with params = 200 OK
echo "TEST [4] hr request with params = 200 OK"
curl -I 'http://0.0.0.0:9292/get_next?callnum=HB172.5%20.F692%202019&start=1' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'
curl -I 'http://0.0.0.0:9292/get_previous?callnum=HB172.5%20.F692%202019&start=1' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'
curl -I 'http://0.0.0.0:9292/get_carousel?callnum=HB172.5%20.F692%202019' -H 'X-Requested-With: XMLHttpRequest' -H 'Accept: text/javascript'
```

[DISCOVERYACCESS-8172]: https://culibrary.atlassian.net/browse/DISCOVERYACCESS-8172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DISCOVERYACCESS-8180]: https://culibrary.atlassian.net/browse/DISCOVERYACCESS-8180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/blacklight_cornell/virtual_browse.rb [hunk 1]
```diff
@@ -154,6 +154,7 @@ def build_class_label(classlabel)
 
   # /get_previous
   def previous_callnumber
+    params.require([:callnum, :start])  # sometimes bots are calling this without required parameters - 400
```
Comment (chrisrlc) [inline lib/blacklight_cornell/virtual_browse.rb:157]: Nice - I'd like to see us be more explicit about what parameters are allowed and required in general more in controller actions (e.g. [strong parameters](https://guides.rubyonrails.org/action_controller_overview.html#strong-parameters). This is a great step in that direction.

File: spec/controllers/catalog_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+  context 'when headers and params are missing' do
+    it 'raises an ParameterMissing error' do
+      expect {
+        get :build_carousel
+      }.to raise_error ActionController::ParameterMissing
+    end
+  end
+
+end
```
Comment (chrisrlc) [inline spec/controllers/catalog_controller_spec.rb:219]: Nice job getting these working controller specs - it sounds like you did some research that suggested request specs might be the recommended way now, was that something you experimented with as well? I saw that the rspec documentation actually recommends creating request specs when testing headers (bottom of https://rspec.info/features/6-0/rspec-rails/controller-specs/). But I think having these controller specs is already a big improvement.

File: spec/controllers/catalog_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,219 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe CatalogController, type: :controller do
+
+  # /get_next
+  describe "GET next_callnumber as xhr request and required params" do
```
Comment (chrisrlc) [inline spec/controllers/catalog_controller_spec.rb:8]: There are a few rspec style guides out there that might help with the organization of these specs. In particular: https://www.betterspecs.org/#contexts and https://github.com/rubocop/rspec-style-guide#use-contexts have recommendations that I try to follow. Contexts can also be nested. I usually go with something like:

```
describe "GET next_callnumber" do
    context "when request is valid" do
        it "returns a 200"
    end

   context "when required parameters are missing" do
       it "returns a 500"
   end
end

describe "GET previous_callnumber"
     context "1"

     context "2"

     context "...."
end

describe "GET build_carousel" ...
```

File: spec/controllers/catalog_controller_spec.rb (hunk 2 above)
Comment (bmckinney) [reply spec/controllers/catalog_controller_spec.rb:219]: yes, i tried request specs but it gets complicated when the api is an external one
once i get a bit more experience i'll try again but as you say, this is a slight improvement

**Pull request merged**